### ISSUE TITLE
Tsbuild watching packagejson is dependent on whether program was built in the invocation

### ISSUE THREAD

**Issue by sheetalkamat** [microsoft/TypeScript#48314]
Because these package json lookups are calculated when program is built.. If the program was already built and then you change package.json file and invoke tsbuild, the change wont be picked by ts build and it will be reported as upto date. 

Also note that ts build uses the infomation from tsconfig to derive if things can be upto date. So we never watch module resolution etc. So this seems one off.
ts build also doesnt mean its incremental build so storing this info in tsbuildinfo is not enough.

_Originally posted by @sheetalkamat in https://github.com/microsoft/TypeScript/pull/44935#r829461596_

**Comment by sheetalkamat** [comment]
cc @weswigham

**Comment by weswigham** [comment]
Yeah, definitely something we'd like to fix; unfortunately the way build mode operates just doesn't lend itself to a way to easily do this without doing exactly what build mode was intending to avoid - parsing every file for (nonrelative) specifiers and resolving them. The existing logic of pulling the data from the program cache is probably also a worthwhile optimization for non-initial builds, even if/when we can find a way to get the list for an initial build.

**Comment by sheetalkamat** [comment]
But it seems like somewhat of inconsistent experience. 

if i did 
```
tsc -b

some change to package.json

tsc -b => // This does not detect change..  and completes as is.. At which point only way to observe change is to do clean compile or delete tsbuildinfo. 
// If we are relying here on change to source file or some such to observe the change why not in watch..
```

The behavior is different vs
```
tsc -b -w which detects the change and does the build.
```

The file watchers are normally polling so they have more cost associated with them.
What's the probability of only changing packge.json vs directory update (like npm install which replaces the file) if that is the case, may be directory watcher is better which is what we use for failed lookups.. 

Having said that i strongly believe, ts build should watch things that are derivable by config file or some heuristic files that can be observed with or without program creation

**Comment by weswigham** [comment]
> The file watchers are normally polling so they have more cost associated with them.
What's the probability of only changing packge.json vs directory update (like npm install which replaces the file) if that is the case, may be directory watcher is better which is what we use for failed lookups..

So in a monorepo setting (which is a primary usecase for build mode, if I recall correctly) only a package file changing without the whole directory structure changing is super common (about as common as people refactoring their file structure or packages I suppose). Package files are essentially simplified source files - they literally have imports in them, people change them as they please (and in ever increasing volume with modern package file features). Outside of that, most node_modules folders are usually installed once and done, and rarely updated.

Do we not use directory watchers wherever possible already? I'd have assumed a single directory watcher in the most-root location relevant to us that we then filter is what we already tried to do if the host supported it. I guess I'm not familiar with the performance characteristics of various file watching strategies, however.

> Having said that i strongly believe, ts build should watch things that are derivable by config file or some heuristic files that can be observed with or without program creation

Watching _every_ possoble package file path in a recursive directory structure is certainly an option I suppose (would be cheap and easy with a directory watcher I imagine). Certainly derivable without program data, and then the packages actually used in program creation can refine that in watch mode to a smaller set of relevant locations.
